### ISSUE TITLE
how to use early stop and lr schedule?

### ISSUE THREAD

**Issue by hust-kevin** [pytorch/ignite#560]
hi, can you give me a example how to use early stop and lr schedule (I want to set new_lr=lr*0.1 after per 10 epoch)

**Comment by vfdev-5** [comment]
Hi @hust-kevin,

the following code does Early Stopping on a monitored validation metric:
```
trainer = ...
evaluator = ...

def score_fn(engine):
    score = engine.state.metrics['Accuracy']
    return score

# Add early stopping with patience of 10 
# if 10 times in a row score_fn gives a result lower than the best result, 
# than the training will be stopped
es_handler = EarlyStopping(patience=10, score_function=score_fn, trainer=trainer)
evaluator.add_event_handler(Events.COMPLETED, es_handler)
```
For a complete example, take a look at : https://github.com/pytorch/ignite/blob/master/examples/notebooks/EfficientNet_Cifar100_finetuning.ipynb

For LR scheduling you have two possibilities:
a) use torch lr schedulling
b) use ignite param schedulling (use with care in case of mixing with NVidia/APEX, https://github.com/pytorch/ignite/issues/545)

Example for a)
```
from torch.optim.lr_scheduler import ExponentialLR

trainer = ...

lr_scheduler = ExponentialLR(optimizer, gamma=0.975)
trainer.add_event_handler(Events.EPOCH_COMPLETED, lambda engine: lr_scheduler.step())
```
https://github.com/pytorch/ignite/blob/master/examples/notebooks/EfficientNet_Cifar100_finetuning.ipynb

Example for b)
```
from ignite.contrib.handlers import PiecewiseLinear

lr = 0.0002
milestones_values = [
    (0, lr),
    (100, lr),
    (200, 0.0)
]
lr_scheduler = PiecewiseLinear(optimizer, param_name='lr', milestones_values=milestones_values)
trainer.add_event_handler(Events.EPOCH_STARTED, lr_scheduler)
```
https://github.com/pytorch/ignite/blob/master/examples/notebooks/CycleGAN.ipynb

HTH

Other links:
- https://pytorch.org/ignite/v0.2.0/contrib/handlers.html#more-on-parameter-scheduling
- https://pytorch.org/ignite/v0.2.0/contrib/handlers.html#more-on-parameter-scheduling

**Comment by hust-kevin** [comment]
@vfdev-5  thanks for your reply, if I want to to print current lr in very iteration, how can I get the current lr?

**Comment by vfdev-5** [comment]
@hust-kevin this can be simpy done checking lr value from optimizer:
```python
@trainer.on(Events.ITERATION_COMPLETED)
def log_lr(engine):
    print(engine.state.iteration, optimizer.param_groups[0]['lr'])
```

For more details how this can be simply done with Tensorboard logger, take a look here :  https://pytorch.org/ignite/v0.2.0/_modules/ignite/contrib/handlers/tensorboard_logger.html#OptimizerParamsHandler

**Comment by hust-kevin** [comment]
thanks
